src/function.py

```diff
@@ -258,11 +262,32 @@ def _generate_payloads(data, split_function):
 
 def _get_license_key(license_key=None):
     """
-    This functions gets New Relic's license key from env vars.
+    This functions gets New Relic's license key from env vars or AWS secrets manager.
     """
+    global cached_license_key
     if license_key:
         return license_key
-    return os.getenv("LICENSE_KEY", "")
+
+    if cached_license_key:
+        return cached_license_key
+
+    # If a secrets manager ARN was specified, use it to grab the license key, otherwise
+    # obtain it from the LICENSE_KEY environment variable.
+    secret_arn = os.getenv("SECRET_KEY_ARN", "")
+    if secret_arn:
+        secrets_client = boto3.client("secretsmanager")
+        secret = secrets_client.get_secret_value(SecretId=secret_arn)
+        cached_license_key = secret.get("SecretString")
+        if _debug_logging_enabled():
+            print(
+                "(got license key from secrets manager, len={})".format(
+                    len(cached_license_key)
+                )
+            )
+    else:
+        cached_license_key = os.getenv("LICENSE_KEY", "")
```

Comment:
Let's use the LICENSE_KEY env var first. If it's set, we use it and are done. Otherwise, we fall back to secrets manager, which has considerable overhead.

In addition, we should move the secret manager fetch to the cold start period, rather than first invocation. 

Finally, I suspect the global state you've introduced here is responsible for the test failures. You'll need to adjust the tests so that they don't have side effects, and aren't dependent on execution order.

Reply:
@MattWhelan You are correct, the global state makes the unit tests challenging. To address that and your other concerns, I'd like to change the code so that `_get_license_key` is called exactly once, from `lambda_handler` (which I believe is the "cold start period" to which you're referring).

That is a much more involved change, as the license key then needs to be passed down to everything that `lambda_handler` calls, but it seems to me like the right approach. Would you agree?